WcaOnRails/spec/structure_sql_spec.rb

```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe "structure.sql" do
+  let!(:structure_sql) do
+    # Load structure.sql excluding the archived phpbb3 tables as well as the Rails internal ones.
+    File.read("#{Rails.root}/db/structure.sql")
```

Comment:
Ok, after some playing around, I've go this:

## Database connection collation and charset

(these methods are defined [here](https://github.com/rails/rails/blob/c33fb580cd00cea476fe211586ca73787e2f0759/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb#L276-L284))

```ruby
irb(main):021:0> ActiveRecord::Base.connection.charset
=> "utf8mb4"
irb(main):022:0> ActiveRecord::Base.connection.collation
=> "utf8mb4_unicode_ci"
```

## Database collation and charset

```ruby
irb(main):020:0> ActiveRecord::Base.connection.select_one("SELECT * FROM information_schema.schemata WHERE schema_name = '#{ActiveRecord::Base.connection_config[:database]}'")
   (0.8ms)  SELECT * FROM information_schema.schemata WHERE schema_name = 'wca_development'
=> {"CATALOG_NAME"=>"def", "SCHEMA_NAME"=>"wca_development", "DEFAULT_CHARACTER_SET_NAME"=>"utf8mb4", "DEFAULT_COLLATION_NAME"=>"utf8mb4_unicode_ci", "SQL_PATH"=>nil}
```

## Table collation and charset

You'll need to iterate over every table (`ActiveRecord::Base.connection.tables`).

(adapted from [this SO answer](https://stackoverflow.com/questions/1049728/how-do-i-see-what-character-set-a-mysql-database-table-column-is))

```ruby
irb(main):030:0> ActiveRecord::Base.connection.select_one("SELECT CCSA.character_set_name, CCSA.COLLATION_NAME FROM information_schema.tables T,        information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` CCSA WHERE CCSA.collation_name = T.table_collation   AND T.table_schema = '#{ActiveRecord::Base.connection_config[:database]}'   AND T.table_name = 'Competitions'")
   (1.6ms)  SELECT CCSA.character_set_name, CCSA.COLLATION_NAME FROM information_schema.tables T,        information_schema.`COLLATION_CHARACTER_SET_APPLICABILITY` CCSA WHERE CCSA.collation_name = T.table_collation   AND T.table_schema = 'wca_development'   AND T.table_name = 'Competitions'
=> {"character_set_name"=>"utf8mb4", "COLLATION_NAME"=>"utf8mb4_general_ci"}
```

## Column collation and charset

You'll need to iterate over every column of every table. You should be able to use the logic in 20170404184332_convert_utf8_to_utf8mb4.rb to do this.

```ruby
irb(main):026:0> ActiveRecord::Base.connection.select_one("SELECT * FROM information_schema.columns WHERE table_schema = '#{ActiveRecord::Base.connection_config[:database]}' AND table_name='Persons' AND column_name='gender'")
   (1.1ms)  SELECT * FROM information_schema.columns WHERE table_schema = 'wca_development' AND table_name='Persons' AND column_name='gender'
=> {"TABLE_CATALOG"=>"def", "TABLE_SCHEMA"=>"wca_development", "TABLE_NAME"=>"Persons", "COLUMN_NAME"=>"gender", "ORDINAL_POSITION"=>5, "COLUMN_DEFAULT"=>"", "IS_NULLABLE"=>"YES", "DATA_TYPE"=>"varchar", "CHARACTER_MAXIMUM_LENGTH"=>1, "CHARACTER_OCTET_LENGTH"=>4, "NUMERIC_PRECISION"=>nil, "NUMERIC_SCALE"=>nil, "DATETIME_PRECISION"=>nil, "CHARACTER_SET_NAME"=>"utf8mb4", "COLLATION_NAME"=>"utf8mb4_general_ci", "COLUMN_TYPE"=>"varchar(1)", "COLUMN_KEY"=>"", "EXTRA"=>"", "PRIVILEGES"=>"select,insert,update,references", "COLUMN_COMMENT"=>""}
```

I'm not actually sure that this is better than what you've done, but if we ever do want to switch from structure.sql back to schema.rb, this test is going to make that switch more difficult.

I'm also worried that the test you've written here could silently start to check nothing if the format of structure.sql changes even very slightly (for example, maybe `DEFAULT CHARSET=utf8mb4` changes to `DEFAULT CHARSET = utf8mb4`.